src/symbol/symbol_layout.js

```diff
@@ -275,9 +276,20 @@ export function performSymbolLayout(bucket: SymbolBucket,
                 if (textJustify === "auto") {
                     textJustify = getAnchorJustification(textAnchor);
                 }
+
+                // text-offset is applied in screen-space in projection.js, where character spacing around curved lines can be
+                // corrected. text-anchor is applied to individual characters, where spacing cannot be corrected.  Thus, *only
+                // for text along lines*, we remove the vertical component of text-anchor and instead pass that value as an
+                // additional (flip-dependent) offset, to be added to text-offset during projection.
+                let adjustedTextAnchor = textAnchor;
+                if (textAlongLine) {
+                    flippedLineOffsetY = getAnchorBaselineShift(textAnchor) * lineHeight;
```

Comment:
I am not quite sure if `lineHeight` is the correct value to multiply with. Since during shaping some line's height may be changed if there is image inside the text, see https://github.com/mapbox/mapbox-gl-js/blob/54615bd65153f1969303098554ce3a076681489d/src/symbol/shaping.js#L742-L744 

So would it better to calculate the `flippedLineOffsetY` after shaping finished by using `shaping.bottom - shaping.top`